File: dart/lib/base_n.dart
library base_N.base_n;

import "dart:convert";
import "dart:typed_data";
import 'dart:math';
import "package:crypto/crypto.dart" show sha256;

int _logint(int v) => (log(v) * 10000).truncate();

final alphabets = Map.unmodifiable({
  2: "01",
  8: "01234567",
  11: "0123456789a",
  16: "0123456789abcdef",
  32: "0123456789ABCDEFGHJKMNPQRSTVWXYZ",
  36: "0123456789abcdefghijklmnopqrstuvwxyz",
  58: "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
  62: "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ",
  64: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/",
  67: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.!~",
});


class CodecDirection {
  final int base1, base2;
  final int log1, log2;

  CodecDirection(this.base1, this.base2)
      : log1 = _logint(base1),
        log2 = _logint(base2);

  int aproximateSize(int size) {
    return 1 + size * log1 ~/ log2;
  }

  int divmod(List<int> digits, int startAt) {
    int remaining = 0;
    for (int i = startAt; i < digits.length; i++) {
      int num = base1 * remaining + digits[i];
      digits[i] = num ~/ base2;
      remaining = num % base2;
    }
    return remaining;
  }
}

var _cached_alphabets = {};

class Alphabet {
  final String key;
  final List<int> _forward;
  final Map<int, int> _inverse;
  final CodecDirection encode_direction;
  final CodecDirection decode_direction;

  Alphabet._internal(text)
      : key = text,
        _forward = List.unmodifiable(text.runes),
        _inverse = Map.unmodifiable(Map.fromIterables(
            text.runes, List.generate(text.length, (i) => i))),
        decode_direction = CodecDirection(text.length, 256),
        encode_direction = CodecDirection(256, text.length);

  factory Alphabet(String text) {
    return _cached_alphabets.putIfAbsent(text, () => Alphabet._internal(text));
  }

  factory Alphabet.predefined(int id) {
    return Alphabet(alphabets[id]);
  }

  List<int> to_digits(String chars) => [for (var c in chars.runes) _inverse[c]];

  String to_chars(List<int> digits) =>
      String.fromCharCodes([for (var d in digits) _forward[d]]);
}

List<int> _sha256x2(List<int> b) =>
    sha256.convert(sha256.convert(b).bytes).bytes;

bool _list_equals(
    List<int> list1, int offset1, List<int> list2, int offset2, int count) {
  for (int i = 0; i < count; i++) {
    if (list1[offset1 + i] != list2[offset2 + i]) {
      return false;
    }
  }
  return true;
}

abstract class BaseN extends Codec<List<int>, String> {
  final Alphabet alphabet;
  DigitsEncoder _encoder;
  DigitsDecoder _decoder;

  List<int> codeDigits(List<int> digits, CodecDirection direction) {
    if (digits.length == 0) return [];
    int leadingZeros = countLeadingZeros(digits);
    int codeSize = direction.aproximateSize(digits.length - leadingZeros);
    Uint8List out = Uint8List(leadingZeros + codeSize);
    int firstNonZero = repackDigits(digits, direction, leadingZeros, out);
    return firstNonZero == leadingZeros
        ? out
        : out.sublist(firstNonZero - leadingZeros);
  }

  int repackDigits(List<int> digits, CodecDirection direction, int leadingZeros,
      Uint8List out);

  String encodeCheck(List<int> bytes) {
    Uint8List buffer = Uint8List(bytes.length + 4);
    buffer.setRange(0, bytes.length, bytes);
    List<int> checksum = _sha256x2(bytes);
    buffer.setRange(bytes.length, buffer.length, checksum);
    return encode(buffer);
  }

  List<int> decodeCheck(String code) {
    List<int> bytes = decode(code);
    int payloadEnd = bytes.length - 4;
    Uint8List buffer = Uint8List.fromList(bytes.sublist(0, payloadEnd));
    List<int> checksum = _sha256x2(buffer);
    if (!_list_equals(checksum, 0, bytes, payloadEnd, 4)) {
      throw FormatException('Checksum does not match');
    }
    return buffer;
  }

  BaseN(Alphabet _alphabet) : alphabet = _alphabet {
    _encoder = DigitsEncoder(this);
    _decoder = DigitsDecoder(this);
  }

  @override
  Converter<List<int>, String> get encoder => _encoder;

  @override
  Converter<String, List<int>> get decoder => _decoder;
}

int countLeadingZeros(List<int> digits) {
  int z = 0;
  while (z < digits.length && digits[z] == 0) z++;
  return z;
}

class DigitsEncoder extends Converter<List<int>, String> {
  final BaseN logic;
  DigitsEncoder(BaseN this.logic);

  @override
  String convert(List<int> bytes) => logic.alphabet.to_chars(
      logic.codeDigits(List.from(bytes), logic.alphabet.encode_direction));
}

class DigitsDecoder extends Converter<String, List<int>> {
  final BaseN logic;
  DigitsDecoder(BaseN this.logic);

  @override
  List<int> convert(String text) => this.logic.codeDigits(
      this.logic.alphabet.to_digits(text), logic.alphabet.decode_direction);
}

class LoopBaseN extends BaseN {
  LoopBaseN(Alphabet alphabet) : super(alphabet);

  @override
  int repackDigits(List<int> digits, CodecDirection direction, int leadingZeros,
      Uint8List out) {
    int startAt = leadingZeros;
    int j = out.length;
    int firstNonZero = j;
    while (startAt < digits.length && leadingZeros < j) {
      int mod = direction.divmod(digits, startAt);
      if (digits[startAt] == 0) startAt++;
      out[--j] = mod;
      if (mod != 0) firstNonZero = j;
    }
    return firstNonZero;
  }
}

class BigIntBaseN extends BaseN {
  BigIntBaseN(Alphabet alphabet) : super(alphabet);

  @override
  int repackDigits(List<int> digits, CodecDirection direction, int leadingZeros,
      Uint8List out) {
    var base1 = BigInt.from(direction.base1);
    var base2 = BigInt.from(direction.base2);
    BigInt acc = digits
        .sublist(leadingZeros)
        .fold(BigInt.zero, (p, n) => p * base1 + BigInt.from(n));
    int j = out.length;
    int firstNonZero = j;
    while (acc > BigInt.zero) {
      int mod = (acc % base2).toInt();
      acc = acc ~/ base2;
      out[--j] = mod;
      if (mod != 0) firstNonZero = j;
    }
    return firstNonZero;
  }
}

File: kotlin/src/main/com/github/walnutgeek/base_n/base_n.kt
package com.github.walnutgeek.base_n

import java.lang.Math.log
import java.math.BigInteger
import java.security.MessageDigest
import kotlin.math.ln

val ALPHABETS = mapOf(
    2 to "01",
    8 to "01234567",
    11 to "0123456789a",
    16 to "0123456789abcdef",
    32 to "0123456789ABCDEFGHJKMNPQRSTVWXYZ",
    36 to "0123456789abcdefghijklmnopqrstuvwxyz",
    58 to "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz",
    62 to "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ",
    64 to "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/",
    67 to "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.!~"
)

private fun sha256x2(input: ByteArray): ByteArray = sha256(sha256(input))

private fun sha256(input: ByteArray) = MessageDigest.getInstance("SHA-256").digest(input)

private fun logInt(v:Int): Int = (ln(v.toDouble()) * 10000).toInt()

class CodecDirection (val fromBase:Int, val toBase :Int) {
    val fromLog = logInt(fromBase)
    val toLog = logInt(toBase)

    fun approximateSize(size:Int):Int = 1 + size * fromLog / toLog

    fun divmod(digits:IntArray, startAt:Int):Int{
        var remaining = 0
        for (i in startAt until digits.size) {
            val num = fromBase * remaining + digits[i]
            digits[i] = num / toBase
            remaining = num % toBase
        }
        return remaining
    }
}


class Alphabet private constructor(val key:String){
    companion object Factory {
        private val cache = mutableMapOf<String,Alphabet>()

        fun predefined(id: Int) : Alphabet {
            return fromString(ALPHABETS[id]!!)
        }

        fun fromString(key : String) : Alphabet {
            return cache[key] ?: Alphabet(key).also{ cache[key] = it }
        }
    }
    val inverse: Map<Char, Int> = mapOf( *key.mapIndexed { i:Int, ch:Char -> ch to i }.toTypedArray())
    val encoder = CodecDirection(256, key.length)
    val decoder = CodecDirection(key.length, 256)

    fun toDigits(chars: String):IntArray = IntArray (chars.length) {
            i-> inverse[chars[i]] ?:
            throw java.lang.IndexOutOfBoundsException("'${chars[i]}' not in alphabet")}

    fun toChars(digits:IntArray ):String = String(CharArray(digits.size) {
            i-> digits[i].let {
                if ( it < key.length )
                    key[it]
                else
                    throw java.lang.IndexOutOfBoundsException("digit:${it} not in alphabet")
            }})

}

tailrec fun countLeadingZeros(digits: IntArray, z: Int = 0): Int {
    return if (z < digits.size && digits[z] == 0) countLeadingZeros(digits, z+1) else z
}

abstract class BaseN {
    abstract val alphabet:Alphabet

    fun encode(bytes:ByteArray):String = alphabet.toChars(
            codeDigits( IntArray(bytes.size){ i-> bytes[i].toInt() and 0xFF}, alphabet.encoder))

    fun decode(text:String):ByteArray = codeDigits(alphabet.toDigits(text), alphabet.decoder).let{
        ByteArray(it.size) {i->it[i].toByte()}}

    fun codeDigits( digits:IntArray,  direction:CodecDirection):IntArray {
        if (digits.isEmpty()) return IntArray(0)
        val leadingZeros = countLeadingZeros(digits)
        val codeSize = direction.approximateSize(digits.size - leadingZeros)
        val out= IntArray(leadingZeros + codeSize)
        val firstNonZero = repackDigits(digits, direction, leadingZeros, out)
        return if (firstNonZero == leadingZeros ) out else out.copyOfRange(firstNonZero - leadingZeros,out.size)
    }

    abstract fun repackDigits( digits:IntArray,  direction:CodecDirection, leadingZeros:Int, out:IntArray):Int

    fun encodeCheck( bytes:ByteArray): String {
        val buffer = ByteArray(bytes.size + 4)
        bytes.copyInto(buffer)
        val checksum = sha256x2(bytes)
        checksum.copyInto(buffer, bytes.size, endIndex = 4)
        return encode(buffer)
    }

    fun decodeCheck(code:String) :ByteArray{
        val bytes = decode(code)
        val payloadEnd = bytes.size - 4
        val buffer = ByteArray(payloadEnd) {i->bytes[i]}
        val checksum = sha256x2(buffer).slice(0..3)
        if (checksum != bytes.slice(payloadEnd until bytes.size) ) {
            throw IllegalStateException("Checksum does not match")
        }
        return buffer
    }

}


class LoopBaseN(override val alphabet:Alphabet) : BaseN() {

    override fun repackDigits(digits:IntArray, direction:CodecDirection, leadingZeros:Int, out:IntArray): Int {
        var startAt = leadingZeros
        var j = out.size
        var firstNonZero = j
        while (startAt < digits.size && leadingZeros < j) {
            val mod = direction.divmod(digits, startAt)
            if (digits[startAt] == 0) startAt++
            out[--j] = mod
            if (mod != 0) firstNonZero = j
        }
        return firstNonZero
    }
}

class BigIntBaseN(override val alphabet:Alphabet) : BaseN() {

    override fun repackDigits(digits:IntArray, direction:CodecDirection, leadingZeros:Int,
                              out:IntArray): Int {
        val fromBase = BigInteger.valueOf(direction.fromBase.toLong())
        val toBase = BigInteger.valueOf(direction.toBase.toLong())
        var acc = digits.slice(leadingZeros until digits.size)
                .fold(BigInteger.ZERO) {p, n -> p * fromBase + BigInteger.valueOf(n.toLong()) }
        var j = out.size
        var firstNonZero = j
        while (acc > BigInteger.ZERO) {
            val (newAcc,bigMod) = acc.divideAndRemainder(toBase)
            val mod = bigMod.toInt()
            acc = newAcc
            out[--j] = mod
            if (mod != 0) firstNonZero = j
        }
        return firstNonZero
    }
}
